apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+                {
+                    TaskCompletionSource<int> signal = new TaskCompletionSource<int>();
+                    s_waitQueue.Enqueue(signal);
+                    await Task.WhenAny(signal.Task, canceledTask).ConfigureAwait(false);
+                    cancellationToken.ThrowIfCancellationRequested();
+                    lock (s_waitSync)
+                    {
+                        if (CurrentActiveSessions < MaximumActiveSessions)
+                        {
+                            s_sessionsCreating++;
+                            return;
+                        }
+                    }
+                }
+                else
+                {
+                    throw new RpcException(new Status(StatusCode.ResourceExhausted, "Number of available Sessions exhausted."));
+                }
+            }
+        }
+
+        private static void EndSessionCreating()
+        {
+            lock (s_waitSync)
+            {
+                s_sessionsCreating--;
+            }
+            SignalAnyWaitingRequests();
+        }
+
+        /// <summary>
+        /// Allocates a session from the pool if possible, or creates a new Spanner Session.
+        /// </summary>
+        /// <param name="spannerClient"></param>
+        /// <param name="project"></param>
+        /// <param name="spannerInstance"></param>
+        /// <param name="spannerDatabase"></param>
+        /// <param name="options"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public static async Task<Session> CreateSessionFromPoolAsync(this SpannerClient spannerClient,
+            string project, string spannerInstance, string spannerDatabase, TransactionOptions options, CancellationToken cancellationToken)
+        {
+            project.ThrowIfNullOrEmpty(nameof(project));
+            spannerInstance.ThrowIfNullOrEmpty(nameof(spannerInstance));
+            spannerDatabase.ThrowIfNullOrEmpty(nameof(spannerDatabase));
+            
+            await StartSessionCreating(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                var sessionPoolKey = new SessionPoolKey {
+                    Client = spannerClient,
+                    Project = project,
+                    Instance = spannerInstance,
+                    Database = spannerDatabase
+                };
+                SessionPoolImpl targetPool = GlobalPool.GetOrAdd(sessionPoolKey, key => new SessionPoolImpl(key));
+
+                var session = await targetPool.AcquireSessionAsync(options, cancellationToken);
+                if (session != null)
+                {
+                    //refresh the mru list which tells us what sessions need to be trimmed from the pool when we want
+                    // to add another poolEntry.
+                    ReSortMru(targetPool);
+                    s_sessionsInUse.AddOrUpdate(session, x => sessionPoolKey, (x, y) => sessionPoolKey);
```

Comment:
and just to make sure I was clear -- the global cap on session count is something the spanner team defined and requested.  They didnt ask for the balanced eviction -- but I figured it was a good idea.  It will turn out useful because like I commented earlier, I will need to add more grpc channels.  Since sessions are tied (for performance reasons) to a specific grpc channel, this organization of pools to client+db will also allow us to keep a somewhat balanced pool between channels within the same db.